design/dsl/validation.go

```diff
@@ -234,6 +234,10 @@ func Required(names ...string) {
 		return
 	}
 
+	if at.Type != nil && at.Type.Kind() == design.ObjectKind {
+		o := at.Type.(design.Object)
+		checkMissingValidationAttribute(names, o)
```

Comment:
This should be done in the expression `Validate` method. The idea is that the `eval` package first runs the DSLs then calls the `Validate` method on all the expressions and if that passes calls `Finalize` next. Sometimes it is necessary to do the check during DSL execution (for example because it would fail otherwise) but here that doesn't seem necessary. You can see the `AttributeExpr` `Validate` method already does that check: https://github.com/goadesign/goa/blob/v2/design/attribute.go#L160-L170.